library(synapser)
library(Seurat)
library(reshape2)
library(stringr)
library(GEOquery)
library(reticulate)
library(HDF5Array)
library(dplyr)
library(readxl)
library(SingleCellExperiment)
library(sageseqr)
library(scDblFinder)

source("Filenames.R")

# Generic helper functions -----------------------------------------------------------

# Assumes you have already authenticated a Synapse login or have a Synapse
# config set up with proper credentials
DownloadFromSynapse <- function(synIDs, downloadLocation) {
  synLogin()
  files <- list()

  for (name in names(synIDs)) {
    files[[name]] <- synGet(synIDs[[name]]$id,
                            version = synIDs[[name]]$version,
                            downloadLocation = downloadLocation)
  }

  return(files)
}

# Standardize cell type names across all data sets.
RemapCelltypeNames <- function(celltypes) {
  # Use only the first 2 characters (reduces the number of possible variations)
  celltypes_mod <- str_to_lower(str_sub(celltypes, end = 2))

  remap <- melt(list(Astrocyte = c("as"),
                     Endothelial = c("en"),
                     Excitatory = c("ex", "gl"), # gl = Glut
                     Inhibitory = c("in", "ga"), # ga = GABA
                     Microglia = c("mg", "mi"),
                     Oligodendrocyte = c("od", "ol"),
                     OPC = c("op"),
                     Pericyte = c("pe"),
                     VLMC = c("vl")))
  rownames(remap) <- remap$value

  return(remap[celltypes_mod, "L1"])
}

# Use the gene conversion list from Step01 to get the mapping between Ensembl ID
# and gene symbol
EnsemblIdToGeneSymbol <- function(gene_list) {
  gene_conversions <- read.csv(file_gene_list)

  rownames(gene_conversions) <- gene_conversions$ensembl_gene_id
  overlap <- intersect(gene_list, gene_conversions$ensembl_gene_id)

  gene_conversions <- gene_conversions[overlap, ]

  genes <- gene_conversions %>%
    dplyr::select(ensembl_gene_id, canonical_symbol, exon_length) %>%
    dplyr::rename(hgnc_symbol = canonical_symbol) %>%
    dplyr::arrange(ensembl_gene_id)
  return(genes)
}

# We have single cell data sets that were aligned to different versions of
# GRCh 38 and have different symbols for the same genes. This function maps
# old symbols to new ones based on the gene list generated by Step01
UpdateGeneSymbols <- function(dataset, gene_list) {
  version <- switch(dataset,
                    "cain" = "symbol_v98",
                    "lau" = "symbol_v93",
                    "leng" = "symbol_v84",
                    "mathys" = "symbol_Mathys", # They provided their mapping as a file
                    "seaRef" = "symbol_seaRef")

  gene_conversions <- read.csv(file_gene_list)
  gene_conversions$original_symbol <- gene_conversions[, version]

  genes <- subset(gene_conversions, original_symbol %in% gene_list)
  rownames(genes) <- make.unique(genes$original_symbol, sep = "/")

  genes <- genes[rownames(genes) %in% gene_list, ] %>%
    dplyr::select(ensembl_gene_id, canonical_symbol, original_symbol) %>%
    dplyr::rename(hgnc_symbol = canonical_symbol) %>%
    dplyr::arrange(ensembl_gene_id)

  return(genes)
}


# Generic functions that call dataset-specific functions ----------------------

DownloadData <- function(dataset, metadata_only = FALSE) {
  files <- switch(dataset,
                  "cain" = DownloadData_Cain(metadata_only),
                  "lau" = DownloadData_Lau(metadata_only),
                  "leng" = DownloadData_Leng(metadata_only),
                  "mathys" = DownloadData_Mathys(metadata_only),
                  "seaRef" = DownloadData_SEARef(metadata_only),
                  "Mayo" = DownloadData_Mayo(metadata_only),
                  "MSBB" = DownloadData_MSBB(metadata_only),
                  "ROSMAP" = DownloadData_ROSMAP(metadata_only))
  return(files)
}

ReadMetadata <- function(dataset, files) {
  metadata <- switch(dataset,
                     "cain" = ReadMetadata_Cain(files),
                     "lau" = ReadMetadata_Lau(files),
                     "leng" = ReadMetadata_Leng(files),
                     "mathys" = ReadMetadata_Mathys(files),
                     "seaRef" = ReadMetadata_SEARef(files),
                     "Mayo" = ReadMetadata_BulkData(dataset, files),
                     "MSBB" = ReadMetadata_BulkData(dataset, files),
                     "ROSMAP" = ReadMetadata_BulkData(dataset, files))
  return(metadata)
}

ReadCovariates <- function(dataset, files) {
  covariates <- switch(dataset,
                       "cain" = ReadMetadata_Cain(files)$covariates,
                       "lau" = ReadCovariates_Lau(files),
                       "leng" = ReadMetadata_Leng(files)$covariates,
                       "mathys" = ReadMetadata_Mathys(files)$covariates,
                       "seaRef" = ReadMetadata_SEARef(files)$covariates,
                       "Mayo" = ReadMetadata_BulkData(dataset, files)$covariates,
                       "MSBB" = ReadMetadata_BulkData(dataset, files)$covariates,
                       "ROSMAP" = ReadMetadata_BulkData(dataset, files)$covariates)
  return(covariates)
}

ReadCounts <- function(dataset, files) {
  counts <- switch(dataset,
                   "cain" = ReadCounts_Cain(files),
                   "lau" = ReadCounts_Lau(files),
                   "leng" = ReadCounts_Leng(files),
                   "mathys" = ReadCounts_Mathys(files),
                   "seaRef" = ReadCounts_SEARef(files),
                   "Mayo" = ReadCounts_BulkData(files),
                   "MSBB" = ReadCounts_BulkData(files),
                   "ROSMAP" = ReadCounts_BulkData(files))
  return(counts)
}


# Custom functions for each data set ------------------------------------------


## Cain, et al., 2020 [preprint] -----------------------------------------------
# https://doi.org/10.1101/2020.12.22.424084

DownloadData_Cain <- function(metadata_only = FALSE) {
  synIDs <- list("metadata" = list(id = "syn23554294", version = 5),
                 "biospecimen_metadata" = list(id = "syn21323366", version = 17),
                 "clinical_metadata" = list(id = "syn3191087", version = 11),
                 "counts" = list(id = "syn23554292", version = 3),
                 "genes" = list(id = "syn23554293", version = 2))

  if (metadata_only) {
    synIDs <- synIDs[1:3]
  }

  files <- DownloadFromSynapse(synIDs, dir_cain_raw)
  return(files)
}

ReadMetadata_Cain <- function(files) {
  metadata <- read.csv(files$metadata$path)

  # Diagnosis is in the specimen ID for each donor
  dx_defs <- c("Cdx1-pAD0" = "Control",
               "Cog1-Path0" = "Control",
               "Cdx1-pAD1" = "Control w/ Pathology",
               "Cog1-Path1" = "Control w/ Pathology",
               "Cdx4-pAD0" = "Non-AD Dementia",
               "Cog4-Path0" = "Non-AD Dementia",
               "Cdx4-pAD1" = "AD",
               "Cog4-Path1" = "AD")

  tmp <- as.data.frame(str_split(metadata$specimenID, "-", simplify = TRUE))
  tmp <- paste(tmp$V4, tmp$V5, sep = "-")

  metadata$diagnosis <- dx_defs[tmp]

  metadata$Cell.Type <- RemapCelltypeNames(metadata$Cell.Type)

  # Fix the 'no.clus' label for sub.cluster
  no_clust <- metadata$sub.cluster == "no.clus"
  metadata$sub.cluster[no_clust] <- metadata$Cell.Type[no_clust]

  # Assign oligodendrocyte subtypes based on the oligodendrocyte 'topic weights'
  # in the metadata. The paper notes that oligo subtypes are more of a spectrum
  # than discrete subtypes, however for our purposes we partition into discrete
  # subtypes by using the highest-weighted 'topic'.
  cols <- paste0("oligodendrocytes.topicweight.", 1:4)
  oligos <- which(metadata$Cell.Type == "Oligodendrocyte")
  olig_types <- apply(metadata[oligos, cols], 1, which.max)
  metadata$sub.cluster[oligos] <- paste0("Oligo.", olig_types)

  metadata <- metadata[, c("cell_name", "specimenID", "diagnosis",
                           "Cell.Type", "sub.cluster")]

  biospecimen <- read.csv(files$biospecimen_metadata$path)
  biospecimen <- subset(biospecimen, specimenID %in% metadata$specimenID) %>%
    select(individualID, specimenID)

  clinical <- read.csv(files$clinical_metadata$path)
  clinical <- subset(clinical, individualID %in% biospecimen$individualID)
  clinical <- merge(clinical, biospecimen, by = "individualID") %>%
    dplyr::rename(sample = specimenID)

  # Fix age fields to be numeric
  clinical$age_death[clinical$age_death == "90+"] <- 90
  clinical$age_death <- as.numeric(clinical$age_death)
  clinical$age_at_visit_max[clinical$age_at_visit_max == "90+"] <- 90
  clinical$age_at_visit_max <- as.numeric(clinical$age_at_visit_max)

  return(list("metadata" = metadata, "covariates" = clinical))
}

ReadCounts_Cain <- function(files) {
  counts <- ReadMtx(mtx = files$counts$path,
                    cells = files$metadata$path,
                    features = files$genes$path,
                    feature.column = 1,
                    skip.cell = 1,
                    skip.feature = 1)

  # The column names and genes don't get read in correctly by ReadMtx because
  # the input csvs aren't formatted as expected for MTX
  colnames(counts) <- str_replace(colnames(counts), ",.*", "")
  rownames(counts) <- str_replace(rownames(counts), ".*,", "")

  return(counts)
}

QC_Cain <- function(seurat) {
  # This data has already been thresholded by counts and doesn't need additional
  # modification.
  return(seurat)
}


## Lau, et al., 2020 -----------------------------------------------------------
# https://doi.org/10.1073/pnas.2008762117

DownloadData_Lau <- function(metadata_only = FALSE) {
  gse <- getGEO("GSE157827", destdir = dir_lau_raw)

  geo_metadata <- pData(phenoData(gse[[1]]))
  file_geo_metadata <- file.path(dir_lau_raw, "geo_metadata.csv")
  write.csv(geo_metadata, file_geo_metadata)

  if (!metadata_only) {
    # For some reason this function call is failing
    geo <- getGEOSuppFiles(GEO = "GSE157827", makeDirectory = FALSE,
                           baseDir = dir_lau_raw)
    untar(rownames(geo)[1], exdir = dir_lau_raw)
  }

  synIDs <- list("clinical_metadata" = list(id = "syn52308080", version = 1))

  files <- DownloadFromSynapse(synIDs, dir_lau_raw)

  files[["geo_metadata"]] <- file_geo_metadata
  return(files)
}

ReadMetadata_Lau <- function(files) {
  # We have to get the barcodes from individual files
  geo_metadata <- read.csv(files$geo_metadata, row.names = 1) %>%
    select(title, diagnosis.ch1) %>%
    dplyr::rename(sample = title, diagnosis = diagnosis.ch1)

  barcode_files <- list.files(path = dir_lau_raw, pattern = "barcodes",
                              full.names = TRUE)
  barcodes <- lapply(barcode_files, function(filename) {
    bc <- read.table(gzfile(filename)) %>% dplyr::rename(barcode = V1)
    file_info <- str_split(filename, pattern = "/", simplify = TRUE)
    file_info <- str_split(file_info[, length(file_info)],
                           pattern = "_",
                           simplify = TRUE)
    bc$sample <- file_info[, 2]
    bc$barcode <- paste(bc$barcode, bc$sample, sep = "_")
    return(bc)
  })

  metadata <- do.call(rbind, barcodes) %>%
    merge(geo_metadata, by = "sample") %>%
    select(barcode, sample, diagnosis)

  # This data set doesn't come with pre-labeled cells
  metadata$broad_class <- NA
  metadata$sub_class <- NA

  clinical <- ReadCovariates_Lau(files)

  return(list("metadata" = metadata, "covariates" = clinical))
}

ReadCovariates_Lau <- function(files) {
  clinical <- readxl::read_excel(files$clinical_metadata$path, sheet = "Patient_info")
  clinical <- as.data.frame(clinical) %>% dplyr::rename(sample = ID)
  return(clinical)
}

ReadCounts_Lau <- function(files) {
  geo_metadata <- read.csv(files$geo_metadata, row.names = 1)
  samples <- rownames(geo_metadata)
  counts_list <- list()

  for (sample in samples) {
    files <- list.files(path = dir_lau_raw,
                        pattern = paste0(sample, "_.*\\.gz"),
                        full.names = TRUE)

    matrix_file <- grep("matrix", files, value = TRUE)
    barcodes_file <- grep("barcodes", files, value = TRUE)
    features_file <- grep("features", files, value = TRUE)

    counts <- ReadMtx(mtx = matrix_file,
                      cells = barcodes_file,
                      features = features_file)
    colnames(counts) <- paste(colnames(counts),
                              geo_metadata[sample, "title"],
                              sep = "_")

    counts_list[[sample]] <- counts
  }

  # Make sure all counts are in the same gene order
  genes <- unique(unlist(lapply(counts_list, rownames)))
  counts_list <- lapply(counts_list, function(X) {
    X[genes, ]
  })

  counts <- do.call(cbind, counts_list)

  return(counts)
}

QC_Lau <- function(seurat) {
  # We need to threshold on total counts / detected genes after removal of doublets
  seurat$high_expression <- seurat$nCount_RNA > 25000
  seurat$low_expression <- seurat$nFeature_RNA < 400

  seurat$pass_QC <- seurat$pass_QC &
    !seurat$high_expression &
    !seurat$low_expression

  return(seurat)
}


## Leng, et al., 2021 ----------------------------------------------------------
# https://doi.org/10.1038/s41593-020-00764-7

# metadata_only is an unused variable for this dataset as the metadata is
# embedded in the counts RDS files
DownloadData_Leng <- function(metadata_only = FALSE) {
  synIDs <- list("counts_ec" = list(id = "syn22722817", version = 1),
                 "counts_sfg" = list(id = "syn22722860", version = 1))
  files <- DownloadFromSynapse(synIDs, dir_leng_raw)
  return(files)
}

ReadMetadata_Leng <- function(files) {
  sce_ec <- readRDS(files$counts_ec$path)
  sce_sfg <- readRDS(files$counts_sfg$path)

  metadata <- rbind(colData(sce_ec), colData(sce_sfg))

  braak <- list("0" = "Control",
                "2" = "Early Pathology",
                "6" = "AD")

  metadata$cell_id <- rownames(metadata)
  metadata$diagnosis <- unlist(braak[metadata$BraakStage])
  metadata$sub_cluster <- str_replace(metadata$clusterAssignment, "EC:|SFG:", "")

  metadata <- as.data.frame(metadata)

  covariates <- metadata %>%
    select(SampleID, PatientID, BrainRegion, BraakStage, SampleBatch, diagnosis) %>%
    distinct() %>%
    dplyr::rename(sample = SampleID)

  metadata <- metadata %>%
    select(cell_id, SampleID, diagnosis, clusterCellType, sub_cluster)

  return(list("metadata" = metadata, "covariates" = covariates))
}

ReadCounts_Leng <- function(files) {
  sce_ec <- readRDS(files[["counts_ec"]]$path)
  sce_sfg <- readRDS(files[["counts_sfg"]]$path)

  # Both counts matrices have the same genes in the same order already
  return(cbind(counts(sce_ec), counts(sce_sfg)))
}

QC_Leng <- function(seurat) {
  # This data has really low counts in general so the cap for high expression
  # is fairly low too
  seurat$high_expression <- seurat$nCount_RNA > 10000
  seurat$pass_QC <- seurat$pass_QC & !seurat$high_expression

  return(seurat)
}


## Mathys, et al., 2019 --------------------------------------------------------
# http://dx.doi.org/10.1038/s41586-019-1195-2

DownloadData_Mathys <- function(metadata_only = FALSE) {
  synIDs <- list("clinical_metadata" = list(id = "syn3191087", version = 11),
                 "cell_metadata" = list(id = "syn18686383", version = 1),
                 "counts" = list(id = "syn18687958", version = 1),
                 "genes" = list(id = "syn18687959", version = 1))

  if (metadata_only) {
    synIDs <- synIDs[1:2]
  }

  files <- DownloadFromSynapse(synIDs, dir_mathys_raw)
  return(files)
}

ReadMetadata_Mathys <- function(files) {
  clinical <- read.csv(files$clinical_metadata$path)
  cellMeta <- read.table(files$cell_metadata$path, sep = "\t", header = TRUE)
  metadata <- merge(cellMeta, clinical,
                    by = "projid", all.x = TRUE, all.y = FALSE)

  # See https://www.synapse.org/#!Synapse:syn3191090 for this information (cogdx)
  diagnosis.codes <- list("1" = "Control",
                          "2" = "MCI",
                          "3" = "MCI",
                          "4" = "AD",
                          "5" = "AD",
                          "6" = "Other")
  diagnosis.codes <- melt(diagnosis.codes)
  colnames(diagnosis.codes) <- c("diagnosis", "cogdx")

  metadata <- merge(metadata, diagnosis.codes, by = "cogdx")

  # Remove cells from the one sample with an "Other" diagnosis
  metadata <- subset(metadata, diagnosis != "Other")

  covariates <- metadata[, c(colnames(clinical), "diagnosis")] %>%
    distinct() %>%
    dplyr::rename(sample = projid)

  # Fix age fields to be numeric
  covariates$age_death[covariates$age_death == "90+"] <- 90
  covariates$age_death <- as.numeric(covariates$age_death)
  covariates$age_at_visit_max[covariates$age_at_visit_max == "90+"] <- 90
  covariates$age_at_visit_max <- as.numeric(covariates$age_at_visit_max)

  metadata <- metadata %>% select(TAG, projid, diagnosis)
  metadata$broad_class <- NA
  metadata$sub_class <- NA

  return(list("metadata" = metadata, "covariates" = covariates))
}

ReadCounts_Mathys <- function(files) {
  counts <- ReadMtx(mtx = files$counts$path,
                    cells = files$cell_metadata$path,
                    features = files$genes$path,
                    feature.column = 1, skip.cell = 1)
  return(counts)
}

QC_Mathys <- function(seurat) {
  seurat$high_expression <- seurat$nCount_RNA > 25000
  seurat$pass_QC <- seurat$pass_QC &
    !seurat$high_expression

  return(seurat)
}


## Seattle Reference Atlas -----------------------------------------------------
# Reference data set (5 donors): https://portal.brain-map.org/atlases-and-data/rnaseq/human-mtg-10x_sea-ad
# Full data set (84 donors): https://portal.brain-map.org/explore/seattle-alzheimers-disease/seattle-alzheimers-disease-brain-cell-atlas-download?edit&language=en
# Metadata: https://www.synapse.org/#!Synapse:syn28256462

# These files are h5ad (AnnData) files. I read the metadata part in with
# reticulate and the "anndata" package, which handles all the categorical
# renaming from the file.
# However, the easiest way to get the data matrix is to use the HDF5Array
# package instead of reticulate. We can't use the R packages that read in
# AnnData (zellkonverter or anndata) because they clobber something from the
# synapser package that makes it not work right.

# metadata_only is an unused variable for this data set, as we need to download
# the data to access the metadata in the anndata file.
DownloadData_SEARef <- function(metadata_only = FALSE) {
  synIDs <- list("individual_metadata" = list(id = "syn31149116", version = 5))
  files <- DownloadFromSynapse(synIDs, downloadLocation = dir_seaad_raw)

  files[["counts"]] <- file_searef_h5

  if (!file.exists(files[["counts"]]) & !metadata_only) { # Don't re-download, this file is large
    download.file(url_searef_h5, destfile = files[["counts"]], method = "curl")
  }

  return(files)
}

ReadMetadata_SEARef <- function(files) {
  donor_metadata <- read.csv(files$individual_metadata$path, row.names = 1)

  ad <- import("anndata")
  adata <- ad$read_h5ad(files$counts, backed = "r")

  metadata <- adata$obs
  metadata <- merge(metadata, donor_metadata,
                    by.x = "external_donor_name_label",
                    by.y = "individualID")

  metadata <- select(metadata, sample_name, external_donor_name_label,
                     diagnosis, class_label, subclass_label)

  covariates <- subset(donor_metadata,
                       individualID %in% metadata$external_donor_name_label) %>%
    dplyr::rename(sample = individualID)

  covariates$ageDeath <- as.numeric(covariates$ageDeath)

  return(list("metadata" = metadata, "covariates" = covariates))
}

ReadCounts_SEARef <- function(files) {
  counts <- H5ADMatrix(files$counts)

  col_names <- as.character(HDF5Array(files$counts,
                                      file.path("obs", "sample_name")))
  dimnames(counts)[[2]] <- col_names
  return(counts)
}


# Generic single cell QC function ----------------------------------------------

QC_SingleCell <- function(metadata, counts, mt_threshold = 0.05, dataset_name, n_cores = 2) {
  # The seaRef data set has already undergone extensive QC, including doublet
  # removal and removal of cells with percent_mito > 0.05%. Therefore we do not
  # remove any additional cells here.
  if (dataset_name == "seaRef") {
    return(counts)
  }

  # Seed for reproducible results with scDblFinder, based on the dataset name
  seed <- sum(as.numeric(charToRaw(dataset_name)))

  stats <- list(total_cells = ncol(counts))

  sce <- SingleCellExperiment(list("counts" = counts), colData = metadata)
  sce <- scuttle::addPerCellQCMetrics(sce)

  # Small amount of initial thresholding to remove empty droplets
  sce <- sce[, sce$detected > 200 & sce$sum > 200]

  stats$low_expression <- ncol(counts) - ncol(sce)

  # If any samples have > 30% of their cells with percent_mito > 0.05, remove
  # them as potentially poor quality samples. The 30% threshold was determined
  # by visually inspecting the distribution of percent_mito for each sample in
  # each dataset and choosing a threshold that removed the most obviously skewed
  # samples in all datasets.
  # TODO: For Lau, this removes 4 AD and 2 NC samples. Lowering to 0.2 removes 2 more AD samples
  # TODO: For Leng, this removes 1 EC sample. Lowering to 0.2 removes 2 SFG samples
  # TODO: For Mathys, this removes 18 samples. Lowering to 0.2 removes 2 more samples.
  mito_stats <- table(sce$sample, sce$percent_mito > mt_threshold)
  mito_stats <- sweep(mito_stats, 1, rowSums(mito_stats), "/")

  if (ncol(mito_stats) == 2) {
    stats$removed_samples <- names(which(mito_stats[, "TRUE"] > 0.3))
    stats$removed_sample_cells <- sum(sce$sample %in% stats$removed_samples)
  } else {
    # This happens with cain, no samples need to be removed
    stats$removed_samples <- c()
    stats$removed_sample_cells <- 0
  }

  if (length(stats$removed_samples) > 0) {
    cat(str_glue("Removing {length(stats$removed_samples)} sample(s) due to ",
                 "sample-wide high mitochondrial gene expression:"),
        paste(stats$removed_samples, collapse = ", "), "\n")
  }

  sce <- sce[, !(sce$sample %in% stats$removed_samples)]

  cat("Finding doublets...\n")

  sce <- scDblFinder::scDblFinder(
    sce,
    clusters = TRUE,
    samples = "sample",
    nfeatures = 4000,
    BPPARAM = BiocParallel::SnowParam(n_cores, RNGseed = seed)
  )

  stats$doublets <- sum(sce$scDblFinder.class == "doublet")
  stats$doublet_pct <- round(stats$doublets / stats$total_cells * 100, digits = 2)
  cat(str_glue("Found {stats$doublets} doublets ({stats$doublet_pct}% of total)."),
      "\n")

  cat("Removing doublet clusters...\n")
  seurat <- as.Seurat(sce, data = "counts")
  rm(sce)
  gc()

  seurat <- RenameAssays(seurat, assay.name = "originalexp", new.assay.name = "RNA")
  seurat <- seurat |>
    NormalizeData() |>
    FindVariableFeatures(nfeatures = 4000) |>
    ScaleData() |>
    RunPCA() |>
    FindNeighbors(dims = 1:20) |> # 20 seems good for all data sets
    FindClusters(resolution = 5)

  seurat$seurat_clusters <- paste0("C", seurat$seurat_clusters)
  dbl_stats <- table(seurat$seurat_clusters, seurat$scDblFinder.class)
  dbl_stats <- sweep(dbl_stats, 1, rowSums(dbl_stats), "/")
  dbl_clusts <- names(which(dbl_stats[, "doublet"] > 0.5))

  cat(str_glue("Removed {length(dbl_clusts)} doublet clusters."), "\n")

  seurat$singlet <- seurat$scDblFinder.class == "singlet"
  seurat$doublet_cluster <- seurat$seurat_clusters %in% dbl_clusts
  seurat$low_mito <- seurat$percent_mito < mt_threshold

  seurat$pass_QC <- seurat$singlet &
    !seurat$doublet_cluster &
    seurat$low_mito

  seurat$high_expression <- FALSE
  seurat$low_expression <- FALSE

  # Do any dataset-specific QC, which may include filtering on high/low
  # expression
  seurat <- switch(dataset_name,
                   "cain" = QC_Cain(seurat),
                   "lau" = QC_Lau(seurat),
                   "leng" = QC_Leng(seurat),
                   "mathys" = QC_Mathys(seurat))

  # Collect some final stats for printout

  # Cells that weren't marked as doublets but are in doublet clusters
  stats$doublet_cluster_cells <- sum(seurat$doublet_cluster & seurat$singlet)

  # Cells remaining with high mitochondrial expression
  stats$mito_cells <- sum(!seurat$low_mito & !seurat$doublet_cluster & seurat$singlet)

  # Cells remaining with low expression
  stats$low_expression <- stats$low_expression +
    sum(seurat$low_mito & !seurat$doublet_cluster &
          seurat$singlet & seurat$low_expression)

  # Cells remaining with high expression
  stats$high_expression <- sum(seurat$low_mito & !seurat$doublet_cluster &
                                 seurat$singlet & seurat$high_expression)

  stats$pct_pass <- round(sum(seurat$pass_QC) / ncol(counts) * 100, digits = 2)

  cat(str_glue(
    "{sum(seurat$pass_QC)} of {ncol(counts)} cells ({stats$pct_pass}%) passed ",
    "QC. Removed:\n",
    ".. {stats$removed_sample_cells} cells from {length(stats$removed_samples)} ",
    "low-quality sample(s)\n",
    ".. {stats$doublets} doublets\n",
    ".. {stats$doublet_cluster_cells} additional cells in doublet clusters\n",
    ".. {stats$low_expression} cells with low expression\n",
    ".. {stats$high_expression} cells with high expression\n",
    ".. {stats$mito_cells} cells with high mitochondrial expression"
  ), "\n")

  # Save stats for examination
  saveRDS(stats, file.path(dir_tmp, str_glue("{dataset_name}_qc.rds")))

  # Only keep cells that passed QC
  return(counts[, colnames(seurat)[seurat$pass_QC]])
}


## Mayo ------------------------------------------------------------------------
# Bulk RNA seq data from the Mayo RNA Seq Study:
# https://adknowledgeportal.synapse.org/Explore/Studies/DetailsPage/StudyDetails?Study=syn5550404
#
# Metadata: https://www.synapse.org/#!Synapse:syn29855549
# Individual metadata: https://www.synapse.org/#!Synapse:syn23277389
# Filtered counts: https://www.synapse.org/#!Synapse:syn27024951
# Biomart gene conversion: https://www.synapse.org/#!Synapse:syn27024953

DownloadData_Mayo <- function(metadata_only = FALSE) {
  synIDs <- list("metadata" = list(id = "syn29855549", version = 2),
                 "clinical_metadata" = list(id = "syn23277389", version = 6),
                 "counts" = list(id = "syn27024951", version = 1))

  if (metadata_only) {
    synIDs <- synIDs[1:2]
  }

  files <- DownloadFromSynapse(synIDs, dir_mayo_raw)
  return(files)
}


## MSBB ------------------------------------------------------------------------
# Bulk RNA seq data from the Mount Sinai Brain Bank Study:
# https://adknowledgeportal.synapse.org/Explore/Studies/DetailsPage/StudyDetails?Study=syn3159438
#
# Metadata: https://www.synapse.org/#!Synapse:syn29855570
# Individual metadata: https://www.synapse.org/#!Synapse:syn6101474
# Filtered counts: https://www.synapse.org/#!Synapse:syn27068754
# Biomart gene conversion: https://www.synapse.org/#!Synapse:syn27068755

DownloadData_MSBB <- function(metadata_only = FALSE) {
  synIDs <- list("metadata" = list(id = "syn29855570", version = 2),
                 "clinical_metadata" = list(id = "syn6101474", version = 9),
                 "counts" = list(id = "syn27068754", version = 1))

  if (metadata_only) {
    synIDs <- synIDs[1:2]
  }

  files <- DownloadFromSynapse(synIDs, dir_msbb_raw)
  return(files)
}


## ROSMAP ----------------------------------------------------------------------
# Bulk RNA seq data from the ROSMAP study:
# https://adknowledgeportal.synapse.org/Explore/Studies/DetailsPage/StudyDetails?Study=syn3219045
#
# Metadata: https://www.synapse.org/#!Synapse:syn29855598
# Individual metadata: https://www.synapse.org/#!Synapse:syn3191087
# Filtered counts: https://www.synapse.org/#!Synapse:syn26967451
# Biomart gene conversion: https://www.synapse.org/#!Synapse:syn26967452

DownloadData_ROSMAP <- function(metadata_only = FALSE) {
  synIDs <- list("metadata" = list(id = "syn29855598", version = 2),
                 "clinical_metadata" = list(id = "syn3191087", version = 11),
                 "counts" = list(id = "syn26967451", version = 1))

  if (metadata_only) {
    synIDs <- synIDs[1:2]
  }

  files <- DownloadFromSynapse(synIDs, dir_rosmap_raw)
  return(files)
}


## Generic bulk functions ------------------------------------------------------
# These functions all work on Mayo, MSBB, and ROSMAP since the files all come
# from the harmonization effort and are in the same format

ReadMetadata_BulkData <- function(dataset, files) {
  metadata <- read.table(files$metadata$path, header = T)

  # Reclassify samples using the criteria from the Diverse Cohorts phenotype
  # harmonization:
  #   AD: Braak >= 4 and [CERAD = probable AD or definite AD, or Thal >= 2 (Mayo only)]
  #   CT: Braak <= 3 and [CERAD = normal or possible AD, or Thal < 2 (Mayo only)]
  #   OTHER: all other samples
  # I expand the OTHER diagnosis into two more groups:
  #   AD_LOW_PATH: Braak <= 3 and [CERAD = probable AD or definite AD, or Thal >= 2 (Mayo only)]
  #   CT_HIGH_PATH: Braak >= 4 and [CERAD = normal or possible AD, or Thal < 2 (Mayo only)]

  clinical <- read.csv(files$clinical_metadata$path) %>%
    subset(individualID %in% metadata$individualID)

  if (dataset == "Mayo") {
    # Mayo samples do not have CERAD values so we use Thal >= 2 for AD. In the
    # event of NA values for either Braak or Thal, the diagnosis already assigned
    # by Mayo contributors will stay as-is. Some PATH_AGE samples are affected
    # by this but since it's unclear what the pathology is we leave that diagnosis
    # as-is for those samples. We also leave the diagnosis of PSP as-is down below.
    renames <- c("Alzheimer Disease" = "AD", "control" = "CT",
                 "pathological aging" = "PATH_AGE",
                 "progressive supranuclear palsy" = "PSP")
    clinical$diagnosis <- renames[clinical$diagnosis]
    clinical$ad_cerad <- clinical$Thal >= 2
    clinical$high_braak <- clinical$Braak >= 4
    clinical$ad_cerad[clinical$diagnosis == "PSP"] <- NA # Forces PSP to be left alone below
  } else if (dataset == "MSBB") {
    # MSBB CERAD values:
    #   1 = Normal
    #   2 = Definite AD
    #   3 = Probable AD
    #   4 = Possible AD
    clinical$ad_cerad <- clinical$CERAD %in% c(2, 3)
    clinical$high_braak <- clinical$Braak >= 4
  } else if (dataset == "ROSMAP") {
    # ROSMAP CERAD values
    #  1 = Definite AD
    #  2 = Probable AD
    #  3 = Possible AD
    #  4 = No AD/Normal
    clinical$ad_cerad <- clinical$ceradsc %in% c(1, 2)
    clinical$high_braak <- clinical$braaksc >= 4
  }

  # Rows where either ad_cerad or high_braak are NA do not get changed by this code
  clinical$diagnosis[clinical$ad_cerad & clinical$high_braak] <- "AD"
  clinical$diagnosis[!clinical$ad_cerad & !clinical$high_braak] <- "CT"
  clinical$diagnosis[!clinical$ad_cerad & clinical$high_braak] <- "CT_HIGH_PATH"
  clinical$diagnosis[clinical$ad_cerad & !clinical$high_braak] <- "AD_LOW_PATH"

  # Keep 'projid' for ROSMAP for comparison with IHC data
  if (dataset == "ROSMAP") {
    clinical <- clinical %>% select(individualID, diagnosis, projid)
  } else {
    clinical <- clinical %>% select(individualID, diagnosis)
  }

  metadata <- metadata %>% select(-diagnosis) %>%
    merge(clinical, by = "individualID") %>%
    subset(RIN >= 5) # Drop samples with RIN < 5

  # Necessary because the column names of the counts matrix get converted this
  # way automatically
  metadata$specimenID <- make.names(metadata$specimenID)

  # MSBB never had sex swapped / resequencing samples removed from the data
  # before processing, so we do that here
  if (dataset == "MSBB") {
    metadata <- RemoveBadSamples_MSBB(metadata)
  }

  covariates <- metadata
  metadata <- metadata[, c("specimenID", "diagnosis", "tissue")]

  return(list("metadata" = metadata, "covariates" = covariates))
}

ReadCounts_BulkData <- function(files, metadata) {
  counts <- read.table(files[["counts"]]$path, header = TRUE, row.names = 1)
  return(counts)
}


# Removes sex-swapped and pre-determined outlier samples from the MSBB dataset.
# These should have been removed during the RNASeq Harmonization pipeline but
# weren't due to an oversight. There are several samples whose X- or
# Y-chromosomal gene expression does not fit with the sex recorded in the
# metadata, indicating potential contamination or mislabeling of those samples.
# To be consistent with how Mayo and ROSMAP were handled in the RNASeq
# Harmonization pipeline, we remove these samples. Note that the Github
# repository linked below specifies to throw out resequenced samples, however if
# they pass QC I leave them in because they might be useful as technical
# replicates when looking at estimates and errors.
# Source: https://github.com/Sage-Bionetworks/ampad-rnaseq-reprocessing/blob/main/code/metadata_preprocessing/msbb_reprocessing.R
RemoveBadSamples_MSBB <- function(metadata) {
  sex_swapped <- c( 'hB_RNA_12901', 'hB_RNA_12934', 'BM_36_296', 'hB_RNA_10622',
                    'hB_RNA_10622_L43C014', 'hB_RNA_10702', 'hB_RNA_10702_E007C014',
                    'hB_RNA_7765', 'hB_RNA_7765_resequenced', 'hB_RNA_7995',
                    'hB_RNA_8015', 'hB_RNA_8025', 'hB_RNA_8025_E007C014',
                    'hB_RNA_8385', 'hB_RNA_8305')
  outliers <- c('hB_RNA_10577', 'hB_RNA_12964', 'hB_RNA_13144', 'hB_RNA_13397',
                'hB_RNA_10567', 'hB_RNA_10577', 'hB_RNA_7855', 'hB_RNA_9005')

  metadata <- subset(metadata, !(specimenID %in% c(sex_swapped, outliers)))

  return(metadata)
}


# Per-batch outlier detection via PCA. Unlike the RNAseq harmonization study,
# we do this on a per-batch (or tissue) basis, because points tend to cluster by
# batch (ROSMAP) and tissue (Mayo, MSBB), and batch/tissue clusters are
# distinctly separated on the PCA plot which skews how large the standard
# deviation is. We did not see much bias toward diagnosis or sex in the PCA,
# either with all points together or separated by batch.
#
# Arguments:
#   covariates = a data.frame of covariates
#   counts = a matrix of counts, with columns corresponding to specimenIDs in
#            the covariates data.frame
#   sd_threshold = how many standard deviations from mean must a sample be to
#                  be considered an outlier
#   do_plot = whether to plot the PCA of counts, colored by outlier status
#
# Returns:
#   a vector containing the sample names of outliers
FindOutliers_BulkData <- function(dataset, covariates, counts, sd_threshold = 4,
                                  do_plot = FALSE) {
  # Define how to split the data into batches for batch-specific outlier
  # detection. On a PCA plot of all samples, Mayo is distinctly split by tissue
  # but not sequencing batch, MSBB samples all group together with little
  # separation by batch or tissue, and ROSMAP is split by batch, with one
  # addition batch split by tissue.
  covariates$batch <- switch(dataset,
                             "Mayo" = covariates$tissue,
                             "MSBB" = 1, # no separation
                             "ROSMAP" = covariates$final_batch)

  # ROSMAP samples cluster by a mixture of batch and tissue. All batches
  # starting with "1" cluster together on a PCA graph, and all batches starting
  # with "3" cluster together as well, so these batches are collapsed. Batch 2_3
  # separates into one cluster of ACC samples and one cluster of DLPFC/PCC
  # samples so it is split in two. Batch 2_9 also splits by tissue the same way.
  # Batches 2_1 and 2_2 are too small to make inferences about outliers but
  # don't cluster closely with other batches, so they are left alone here and
  # will be skipped (no samples marked as outliers).
  if (dataset == "ROSMAP") {
    covariates$batch[startsWith(covariates$final_batch, "1")] <- "1"
    covariates$batch[startsWith(covariates$final_batch, "3")] <- "3"

    covariates$batch[covariates$batch == "2_3" & covariates$tissue == "ACC"] <- "2_3_ACC"
    covariates$batch[covariates$batch == "2_3" & covariates$tissue != "ACC"] <- "2_3_DLPFC_PCC"
    covariates$batch[covariates$batch == "2_9" & covariates$tissue == "ACC"] <- "2_9_ACC"
    covariates$batch[covariates$batch == "2_9" & covariates$tissue != "ACC"] <- "2_9_DLPFC_PCC"
  }

  # Ensure covariates are in the same order as counts
  rownames(covariates) <- covariates$specimenID
  covariates <- covariates[colnames(counts), ]

  # Find outliers by batch
  outliers <- lapply(unique(covariates$batch), function(batch) {
    batch_samples <- covariates$batch == batch

    if (sum(batch_samples) > 20) {
      # Use sagesqr's function
      res <- sageseqr::identify_outliers(counts[, batch_samples],
                                         covariates[batch_samples, ],
                                         color = "diagnosis",
                                         shape = "tissue", size = "batch",
                                         z = sd_threshold)

      if (do_plot) {
        print(res$plot)
      }

      return(res$outliers)
    }
    return(NULL)
  })

  return(unlist(outliers))
}
